### PR TITLE
Python 3 compatibility

### PR DESCRIPTION
**Pull request by oprypin**
This pull request makes _teuthology_ compatible with Python 3.x while keeping compatibility with Python 2.x.

The most controversial part of this pull request is the [compat.py](https://github.com/BlaXpirit/teuthology/blob/wip-py3/teuthology/compat.py) module and its uses. The reason that it is needed is simple: throughout the _teuthology_ codebase bytestrings are used for text data. It is especially important to not change the existing behavior on Python 2 to Unicode strings due to assumptions outside of _teuthology_, like _ceph-qa-suite_. That's why on Python 2 the _stringify_ function leaves the bytestrings alone, and on Python 3 decodes them into Unicode strings. Same for `StringIO` - on Python 2 it's still `cStringIO` (byte-based) and on Python 3 it's `io.StringIO` (Unicode-based)

Note that I couldn't properly test these changes because _ceph-qa-suite_ is Python 2.x only at the moment. However, I made sure that unittests pass on both Python 2.x and Python 3.x, and that the number-wise stats for the unittests (passed/error/warning etc) are exactly the same as on the master branch.

Additionally, I ran a few actual tests with _teuthology_, again, in 3 configurations: _master_+_python2_, _wip-py3_+_python2_, _wip-py3_+_python3_. Details and logs [**here**](https://gist.github.com/330fe8639e75d54fba7b7dfb461bb605).
Basically, the simple _interactive_ "test" works well, the _teuthology_ test works the same in all configurations, and the _rados/cephtool_ test fails only on Python 3 because it is not compatible with Python 3.

### CODE REVIEW

**Comment by BlaXpirit** [comment]
I dropped the commit that adds *py34* to *tox* so it can pass the check.

There is a problem that the test environment does not have header files for Python 3 installed.

**Comment by jdurgin** [comment]
@BlaXpirit @zmc pointed out that the test script uses './bootstrap install' to get the necessary packages - could you add whatever is needed for py3 there?

**Comment by BlaXpirit** [comment]
@jdurgin Well, it's not as simple as that, because Python 3 development packages are not required to use teuthology with Python 2.

Extending the script to properly support Python 3 separately seems complex: first, it would have to be an additional argument to the script (not entirely clear how to accept separate arguments), then separate packages per version / per OS (especially fun on CentOS).

And in the end the CI job would have to be modified to call it with that argument anyway.

**Comment by jdurgin** [comment]
@BlaXpirit I was thinking of not making it optional - just checking for/installing both py2 and py3 packages, so tox works with both envs by default

**Comment by BlaXpirit** [comment]
@jdurgin, don't forget about users. Python 3 is not available to install by default on CentOS!

**Comment by jdurgin** [comment]
@BlaXpirit users of teuthology are ceph devs - not so worried about extra dependencies, just need to make sure bootstrap supports them, and if needed add additional docs on extra repo setup

Now that we've got py3 ceph packages, ceph devs will have those installed anyway, so it shouldn't be much of an issue

File: teuthology/beanstalk.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 from collections import OrderedDict
 
+import yaml
+try:
+    import beanstalkc
+except ImportError:
+    import pystalkd.Beanstalkd as beanstalkc
+
 from .config import config
 from . import report
```
Comment (jdurgin) [inline teuthology/beanstalk.py]: did you try various teuthology-queue operations? Wondering if there's any more adjustment needed for pystalkd

File: teuthology/run.py [hunk 2]
```diff
@@ -223,7 +223,7 @@ def get_initial_tasks(lock, config, machine_type):
         init_tasks.extend([
             {'pcp': None},
             {'selinux': None},
-            {'ansible.cephlab': None},
+            #{'ansible.cephlab': None},
```
Comment (jdurgin) [inline teuthology/run.py]: will need to remove this commit before merging

File: teuthology/suite/matrix.py [hunk 3]
```diff
@@ -161,7 +161,7 @@ def _index(self, i, submats):
         off = (i // clen) % cycles
 
         def combine(r, s=frozenset()):
-            if type(r) is frozenset:
+            if isinstance(r, frozenset):
```
Comment (jdurgin) [inline teuthology/suite/matrix.py]: nice catch

File: teuthology/beanstalk.py (hunk 1 above)
Comment (BlaXpirit) [inline teuthology/beanstalk.py]: I have not tried any *teuthology-queue* operations.

File: teuthology/run.py (hunk 2 above)
Comment (BlaXpirit) [inline teuthology/run.py]: OK but doesn't it prevent teuthology from working on Python 3?

**Comment by vasukulkarni** [comment]
@rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?

**Comment by vasukulkarni** [comment]
test this please

**Comment by kshtsk** [comment]
> @rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?

@vasukulkarni I actually in the middle of making teuthology code py2/py3 compatible taking code gradually from it, this PR is going to  be outdated as far as I finish the effort and close this PR. In other words, no-one should resolve conflicts within this PR, since the conflicts will increase conflicts.

**Comment by vasukulkarni** [comment]
> > @rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?
> 
> @vasukulkarni I actually in the middle of making teuthology code py2/py3 compatible taking code gradually from it, this PR is going to be outdated as far as I finish the effort and close this PR. In other words, no-one should resolve conflicts within this PR, since the conflicts will increase in amount.

That is good to know, but why have py2 support? Its EOL anyway, but doesn't hurt.

**Comment by kshtsk** [comment]
> > > @rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?
> > 
> > 
> > @vasukulkarni I actually in the middle of making teuthology code py2/py3 compatible taking code gradually from it, this PR is going to be outdated as far as I finish the effort and close this PR. In other words, no-one should resolve conflicts within this PR, since the conflicts will increase in amount.
> 
> That is good to know, but why have py2 support? Its EOL anyway, but doesn't hurt.

Just because we cannot switch everything to py3 at once, we need to keep it working on py2 until all ceph code base be compatible, so the work is staged in following steps:
1) make code compatible py2/py3 (still running infra on py2) (teuthology + ceph release branches)
2) make code runable on pure py3
3) get rid off py2 after stabilization period in infra
4) wipe py2 trace out of the code base

**Comment by vasukulkarni** [comment]
> > > > @rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?
> > > 
> > > 
> > > @vasukulkarni I actually in the middle of making teuthology code py2/py3 compatible taking code gradually from it, this PR is going to be outdated as far as I finish the effort and close this PR. In other words, no-one should resolve conflicts within this PR, since the conflicts will increase in amount.
> > 
> > 
> > That is good to know, but why have py2 support? Its EOL anyway, but doesn't hurt.
> 
> Just because we cannot switch everything to py3 at once, we need to keep it working on py2 until all ceph code base be compatible, so the work is staged in following steps:
> 
> 1. make code compatible py2/py3 (still running infra on py2) (teuthology + ceph release branches)
> 2. make code runable on pure py3
> 3. get rid off py2 after stabilization period in infra
> 4. wipe py2 trace out of the code base

I was talking only about the teuthology code, I think the ceph part would be in ceph repo, but afaik I dont think there is a test code in teuthology.

**Comment by kshtsk** [comment]
> > > > > @rakeshgm @kshtsk @racpatel @zmc there is quite a bit of work here on moving to python3 ? I think one can resolve conflicts, tests and get this merged? WDYT ?
> > > > 
> > > > 
> > > > @vasukulkarni I actually in the middle of making teuthology code py2/py3 compatible taking code gradually from it, this PR is going to be outdated as far as I finish the effort and close this PR. In other words, no-one should resolve conflicts within this PR, since the conflicts will increase in amount.
> > > 
> > > 
> > > That is good to know, but why have py2 support? Its EOL anyway, but doesn't hurt.
> > 
> > 
> > Just because we cannot switch everything to py3 at once, we need to keep it working on py2 until all ceph code base be compatible, so the work is staged in following steps:
> > 
> > 1. make code compatible py2/py3 (still running infra on py2) (teuthology + ceph release branches)
> > 2. make code runable on pure py3
> > 3. get rid off py2 after stabilization period in infra
> > 4. wipe py2 trace out of the code base
> 
> I was talking only about the teuthology code, I think the ceph part would be in ceph repo, but afaik I dont think there is a test code in teuthology.

right the test code is in the ceph repo, but you can't run teuthology on py3 without having ceph  code ported to py3 ;-)

**Comment by vasukulkarni** [comment]
> > I was talking only about the teuthology code, I think the ceph part would be in ceph repo, but afaik I dont think there is a test code in teuthology.
> 
> right the test code is in the ceph repo, but you can't run teuthology on py3 without having ceph code ported to py3 ;-)

Can you tell me the ETA for your PR for python3 support? I might be missing something but ceph repo should be independent work, it should really affect tests in ceph repo. (just the bootstrap and all the parts in teuthology/beanstalkd are python3 ready is what we need in teuthology) and the python2 support is for remote nodes that run work units and is not affected.

**Comment by kshtsk** [comment]
> Can you tell me the ETA for your PR for python3 support? I might be missing something but ceph repo should be independent work, it should really affect tests in ceph repo. (just the bootstrap and all the parts in teuthology/beanstalkd are python3 ready is what we need in teuthology) and the python2 support is for remote nodes that run work units and is not affected.

There is not a single PR, there is a set of PRs which are incorporated into master one-by-one, in order to not break the infrastructure, and minimize teuthology server out of service. Ceph repo can not be independent since qa/tasks are using teuthology modules at source level, and they are not running remotely, test workers are run locally by the worker. Not sure I understand your comment regarding `teuthology/beanstalkd`. Have you tested in on py3 at least once? I never saw it could even run without syntax errors.

**Comment by kshtsk** [comment]
@vasukulkarni not sure I follow your point, would you like to take over the py3 porting effort?

**Comment by vasukulkarni** [comment]
> @vasukulkarni not sure I follow your point, would you like to take over the py3 porting effort?

If you have PR in progress, we dont want to take over, but if its not ready yet then I think there is urgency to get it working now.  I think we can syncup once in Wednesday's meeting so that I can explain why I am thinking ceph repo source files shouldn't be affected and we can fix the teuthology part (based off the work done here) faster.

**Comment by kshtsk** [comment]
So far so good, the time's come and we can close this PR in favor of #1362 since we were able to run successfully ceph tests against master branch on py3 deployed teuthology environment.
